File: ios/CoChat/Repositories/RealtimeStore.swift
import Foundation

/// Swift tuples aren't Equatable even when every element is, and SwiftUI's
/// `.onChange(of:)` requires Equatable — so these are plain structs, not
/// tuples, purely so views can observe them.
struct IncomingMessage: Equatable {
    let chatId: String
    let message: Message
}

struct GroupUpdate: Equatable {
    let chat: GroupInfo
    let action: String
}

@MainActor
final class RealtimeStore: ObservableObject {
    @Published private(set) var chats: [ChatSummary] = []
    @Published private(set) var presence: [String: String] = [:]
    @Published var notifications: [AppNotification] = []
    @Published private(set) var typing: [String: Set<String>] = [:]

    /// Latest live message per chat, for open threads to append without a refetch.
    @Published private(set) var lastIncomingMessage: IncomingMessage?
    @Published private(set) var lastGroupUpdate: GroupUpdate?

    var unreadNotificationCount: Int { notifications.filter { !$0.read }.count }
    var totalUnreadChats: Int { chats.reduce(0) { $0 + $1.unreadCount } }

    private let socket: AppSocketManager
    private let chatsRepo: ChatsRepository
    private let notificationsRepo: NotificationsRepository
    private let tokenStore: TokenStore
    private var started = false

    init(socket: AppSocketManager, chatsRepo: ChatsRepository, notificationsRepo: NotificationsRepository, tokenStore: TokenStore) {
        self.socket = socket
        self.chatsRepo = chatsRepo
        self.notificationsRepo = notificationsRepo
        self.tokenStore = tokenStore
    }

    func start() async {
        guard !started, let token = tokenStore.accessToken else { return }
        started = true

        // Assigned before connect() — the callback closure is live the instant
        // it's set, so the server's own presence:update on connect can't race
        // ahead of a listener the way an async Flow subscription could.
        socket.onEvent = { [weak self] event in
            Task { @MainActor in self?.handle(event) }
        }
        socket.connect(token: token)

        async let fetchedChats = chatsRepo.list()
        async let fetchedNotifications = notificationsRepo.list()
        if let chatsResult = try? await fetchedChats { chats = chatsResult }
        if let notificationsResult = try? await fetchedNotifications { notifications = notificationsResult }

        var snapshot: [String: String] = [:]
        for chat in chats {
            if let peer = chat.peer { snapshot[peer.id] = peer.status }
        }
        presence = snapshot.merging(presence) { _, live in live }
    }

    func stop() {
        socket.disconnect()
        started = false
        chats = []
        presence = [:]
        notifications = []
    }

    func emitTypingStart(chatId: String) { socket.emitTypingStart(chatId: chatId) }
    func emitTypingStop(chatId: String) { socket.emitTypingStop(chatId: chatId) }

    func markChatReadLocally(_ chatId: String) {
        chats = chats.map { chat in
            var updated = chat
            if updated.id == chatId { updated.unreadCount = 0 }
            return updated
        }
    }

    /// Inserts a freshly-created chat (e.g. from "Message" on a team member)
    /// so it's visible immediately, without waiting on a socket event or refetch.
    func upsertChat(_ chat: ChatSummary) {
        if let idx = chats.firstIndex(where: { $0.id == chat.id }) {
            chats[idx] = chat
        } else {
            chats.insert(chat, at: 0)
        }
    }

    func replaceChatsForNewGroup() {
        Task {
            if let list = try? await chatsRepo.list() { chats = list }
        }
    }

    private func handle(_ event: AppSocketEvent) {
        switch event {
        case .presenceUpdate(let userId, let status):
            presence[userId] = status
        case .messageNew(let message, let chatId):
            lastIncomingMessage = IncomingMessage(chatId: chatId, message: message)
        case .chatUpdated(let chat):
            if let idx = chats.firstIndex(where: { $0.id == chat.id }) {
                chats[idx] = chat
            } else {
                chats.insert(chat, at: 0)
            }
            chats.sort { $0.updatedAt > $1.updatedAt }
        case .groupUpdated(let chat, let action):
            lastGroupUpdate = GroupUpdate(chat: chat, action: action)
            if action == "created" { replaceChatsForNewGroup() }
        case .notificationNew(let notification):
            notifications.insert(notification, at: 0)
        case .typingUpdate(let chatId, let userId, let isTyping):
            var set = typing[chatId] ?? []
            if isTyping { set.insert(userId) } else { set.remove(userId) }
            typing[chatId] = set
        case .chatRead:
            break
        }
    }
}

File: web/src/app/core/realtime.store.ts
import { Injectable, computed, signal } from '@angular/core';
import { AuthService } from './auth.service';
import { ChatsService } from './chats.service';
import { NotificationsService } from './notifications.service';
import { SocketService } from './socket.service';
import { AppNotification, ChatSummary, GroupInfo, Message, PresenceStatus } from './models';

interface TypingState {
  chatId: string;
  userId: string;
  isTyping: boolean;
}

@Injectable({ providedIn: 'root' })
export class RealtimeStore {
  readonly chats = signal<ChatSummary[]>([]);
  readonly presence = signal<Record<string, PresenceStatus>>({});
  readonly notifications = signal<AppNotification[]>([]);
  readonly typing = signal<Record<string, Set<string>>>({});
  readonly unreadNotificationCount = computed(() => this.notifications().filter((n) => !n.read).length);
  readonly totalUnreadChats = computed(() => this.chats().reduce((sum, c) => sum + c.unreadCount, 0));

  /** Emits the latest live message per chat so open threads can append without a refetch. */
  readonly lastIncomingMessage = signal<{ chatId: string; message: Message } | null>(null);
  readonly lastGroupUpdate = signal<{ chat: GroupInfo; action: string } | null>(null);

  private started = false;

  constructor(
    private socket: SocketService,
    private auth: AuthService,
    private chatsApi: ChatsService,
    private notificationsApi: NotificationsService
  ) {}

  async start(): Promise<void> {
    if (this.started) return;
    this.started = true;

    const token = this.auth.accessToken;
    if (!token) return;
    this.socket.connect(token);

    // Subscribe before awaiting anything else — the server broadcasts this
    // socket's own presence:update the moment it connects, so registering
    // listeners after an `await` risks missing events that arrive in between.
    this.socket.on<{ userId: string; status: PresenceStatus }>('presence:update').subscribe(({ userId, status }) => {
      this.presence.update((map) => ({ ...map, [userId]: status }));
    });

    this.socket.on<{ message: Message; chatId: string }>('message:new').subscribe(({ message, chatId }) => {
      this.lastIncomingMessage.set({ chatId, message });
    });

    this.socket.on<{ chat: ChatSummary }>('chat:updated').subscribe(({ chat }) => {
      this.chats.update((list) => {
        const idx = list.findIndex((c) => c.id === chat.id);
        const next = idx >= 0 ? [...list.slice(0, idx), chat, ...list.slice(idx + 1)] : [chat, ...list];
        return next.sort((a, b) => new Date(b.updatedAt).getTime() - new Date(a.updatedAt).getTime());
      });
    });

    this.socket.on<{ chat: GroupInfo; action: string }>('group:updated').subscribe(({ chat, action }) => {
      this.lastGroupUpdate.set({ chat, action });
      if (action === 'created') {
        this.chatsApi.list().then((chats) => this.chats.set(chats));
      }
    });

    this.socket.on<AppNotification>('notification:new').subscribe((notification) => {
      this.notifications.update((list) => [notification, ...list]);
    });

    this.socket.on<TypingState>('typing:update').subscribe(({ chatId, userId, isTyping }) => {
      this.typing.update((map) => {
        const set = new Set(map[chatId] ?? []);
        if (isTyping) set.add(userId);
        else set.delete(userId);
        return { ...map, [chatId]: set };
      });
    });

    const [chats, notifications] = await Promise.all([this.chatsApi.list(), this.notificationsApi.list()]);
    this.chats.set(chats);
    this.notifications.set(notifications);

    const presenceSnapshot: Record<string, PresenceStatus> = {};
    for (const chat of chats) {
      if (chat.peer) presenceSnapshot[chat.peer.id] = chat.peer.status;
    }
    // Merge under whatever presence:update events already arrived via the socket
    // (those are newer than this REST snapshot) rather than overwriting them.
    this.presence.update((live) => ({ ...presenceSnapshot, ...live }));
  }

  markChatReadLocally(chatId: string) {
    this.chats.update((list) => list.map((c) => (c.id === chatId ? { ...c, unreadCount: 0 } : c)));
  }

  stop(): void {
    this.socket.disconnect();
    this.started = false;
    this.chats.set([]);
    this.presence.set({});
    this.notifications.set([]);
  }
}
